Enable build of ARM64 installers in ASP.NET Core

hoyosjs opened the pull request:
Enable ARM64 installers build.

- Changes WiX toolset used to 3.14 to support ARM64
- Generates targeting pack from the x86/x64 leg, as it gets produced using a zip that gets generated there.
- The ARM leg now produces all the necessary msi's, exe, and wixlib needed for the installer to generate a bundle.

hoyosjs commented on NuGet.config:
```diff
@@ -11,6 +11,7 @@
     <add key="roslyn-tools" value="https://dotnet.myget.org/F/roslyn-tools/api/v3/index.json" />
     <!-- Used for the SiteExtension 3.1 bits that are included in the 5.0 build -->
     <add key="dotnet31-transport" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet3.1-transport/nuget/v3/index.json" />
+    <add key="general-testing" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/general-testing/nuget/v3/index.json" />
```
I added this feed for testing. The main issue was that WiX doesn't have a package for the toolset that supports ARM64. Other repositories don't use `wixproj`, so we didn't need it before this. I don't know what feed this should go to, or if this is even a good idea.

hoyosjs replied:
Yeah, this was to verify this won't break in CI. I'm asking for some guidance here.

hoyosjs replied:
@dougbu I got the green light to move this as is to `dotnet-eng`. I've reverted all changes to `NuGet.config`

hoyosjs commented on eng/Build.props:
```diff
@@ -64,6 +64,9 @@
         <ProjectToBuild Include="$(RepoRoot)src\Installers\Windows\TargetingPack\TargetingPack.wixproj" AdditionalProperties="Platform=x86" />
+        <!-- This really shouldn't be here, but instead of harvesting, this looks . Producing it in each leg ends up with multiple targeting packs
+        getting produced and the BAR will reject the build. Centralize building the targeting pack in the x86/x64 leg. -->
+        <ProjectToBuild Include="$(RepoRoot)src\Installers\Windows\TargetingPack\TargetingPack.wixproj" AdditionalProperties="Platform=arm64" />
```
This was the other sort of... not-so-clean thing that was necessary. The right solution here is to stop using . This was more straightforward and felt less risky (given this change needs to go to 5.0).

hoyosjs commented:
Official test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=800872&view=results

Further validation: 
- Confirmed files are R2R, signed files.
- Now checking registry keys and signatures of the bundles. 

cc: @joeloff, @tommcdon

Pilchie commented:
Do we want to target this to `release/5.0-rc2`?

hoyosjs commented:
@pilchie, yes. I will port once this gets reviewed and goes in.

joeloff commented:
@Pilchie if we did, we might be able to unblock the SDK and get that part completed for RC2 as well, but it would be tight. I'll have a chat with other SDK folks. @marcpopMSFT as an FYI

tommcdon commented:
FYI @richlander

dougbu commented on src/Installers/Windows/SharedFrameworkBundle/SharedFrameworkBundle.wixproj:
```diff
@@ -29,20 +29,35 @@
+          <DoNotHarvest>True</DoNotHarvest>
+        </ProjectReference>
+      </ItemGroup>
+    </Otherwise>
```
Are we producing a completely separate shared Fx bundle for ARM64 because the platform-specific framework is created in a different job❔ Could the extra bundle confuse customers❔

hoyosjs replied:
After discussion, we decided it doesn't make sense to ship an exe bundle with only the ARM64 installer. The x86+arm64 bundle doesn't provide enough value, and we've deferred this for 6.0 as this is a PR to port for RC2

dougbu commented:
I can't approve this until the `general-testing` feed is removed. @hoyosjs you've addressed my other comments well. (nit: "only" isn't spelled "inly" even in comments.)

Also suggest @joeloff or someone else w/ more Wix knowledge have a look at these changes. They follow the existing patterns but I may be missing something.

dougbu reviewed (approved):
I'm fine with this but would really appreciate additional eyes on it @joeloff @dotnet/aspnet-build or anyone w/ more WiX experience than me, please chime in.

joeloff commented:
Looks good. I'm guessing that the hosting bundle will remain x86/x64 only for the time being. Unlike the other installers, the hosting bundle prefers to install all architectures by default and I think there will be a clash since for x64, there aren't new properties defined yet (at least none that I can see in Windows).

My big concern is the wixlib because we rely on that to share the package chaining across the shared framework, hosting and SDK installer bundles.

hoyosjs commented:
@joeloff any verification I can do to verify? I have official runs so I have signed installers and can test anything that's needed. However, this is becoming time sensitive.

joeloff commented:
@hoyosjs if the have the signed MSI, you can try them out on a VM. You can also double check the Product/Upgrade code for the x64/arm64 MSIs just to be sure they are different (you can use ORCA for that). The build seeds the GUIDs using the Platform property, so they should be different.

hoyosjs commented:
@joeloff I've tested this:

![image](https://user-images.githubusercontent.com/19413848/92648659-9cf0c200-f29e-11ea-9ff3-e346b79f6883.png)

> My big concern is the wixlib because we rely on that to share the package chaining across the shared framework, hosting and SDK installer bundles.

Do you have a particular concern I can check against?

hoyosjs commented:
Issues I keep hitting are https://github.com/dotnet/aspnetcore/issues/25623 and a gradle issue solved by https://github.com/dotnet/aspnetcore/pull/25707. Closing and reopening to create a rebase.

hoyosjs commented:
@dougbu I've tried rebasing but I keep getting red. Looking around I see you had a comment describing all the issues I've seen. However they seem to have gotten fixes upstream, but they keep failing. Any advice?

> * #25651 hit another flaky Razor SDK test (`BundlesScopedCssFiles_UpdatesBundleWhenContentsChange`)   
>   * this test failure impacted most, if not all, of these PRs
> * #25652 hit the PDB duplication build problem in both aspnetcore-ci and aspnetcore-quarantined-pr
> * #25653 hit the `gradlew` build problem @halter73
> * #25654 hit the PDB duplication build problem but only in the aspnetcore-quarantined-pr pipeline

https://github.com/dotnet/aspnetcore/issues/25293#issuecomment-687691562

dougbu commented:
@hoyosjs the problem here is you're targeting the master branch and all the fixes mentioned have been checked into release/5.0-rc2. Need to either wait for #25650 to merge or target release/5.0-rc2 instead. Recommend doing the second because time for RC2 is getting short.

hoyosjs committed:
Enable ARM64 installers build.

Changes WiX toolset used to 3.14 to support ARM64
Generates targeting pack from the x86/x64 leg, as it gets produced using a zip that gets generated there.
The ARM64 leg now produces all the necessary msi's, exe, and wixlib needed for the installer to generate a bundle.

hoyosjs commented:
@dougbu I didn't know there was an automated FI from RC2 to master. I've retargeted this to RC2.

hoyosjs commented:
Feel free to merge once it's green

@Pilchie can this please be considered for RC2 for the ARM64 installer?

dougbu commented:
All checks 🟢 Waiting on your word @Pilchie&hellip;

Pilchie commented:
Just wanted to double check with Tactics - we're good to merge this now.

Pilchie commented:
Approved for .NET 5 RC2.

dougbu commented:
Thanks very much @hoyosjs❕
